aiida_quantumespresso/workflows/pw/relax.py

```diff
@@ -170,41 +214,15 @@ def run_relax(self):
         """Run the `PwBaseWorkChain` to run a relax `PwCalculation`."""
         self.ctx.iteration += 1
 
-        inputs = AttributeDict(self.exposed_inputs(PwBaseWorkChain, namespace='base'))
+        inputs = self.ctx.relax_inputs
         inputs.pw.structure = self.ctx.current_structure
-        inputs.pw.parameters = inputs.pw.parameters.get_dict()
 
-        inputs.pw.parameters.setdefault('CELL', {})
         inputs.pw.parameters.setdefault('CONTROL', {})
         inputs.pw.parameters['CONTROL']['restart_mode'] = 'from_scratch'
 
-        if 'relaxation_scheme' in self.inputs:
-            if self.inputs.relaxation_scheme.value == 'relax':
-                relax_type = RelaxType.ATOMS
-            elif self.inputs.relaxation_scheme.value == 'vc-relax':
-                relax_type = RelaxType.ATOMS_CELL
-            else:
-                raise ValueError('unsupported value for the `relaxation_scheme` input.')
-        else:
-            relax_type = RelaxType(self.inputs.relax_type)
-
-        if relax_type in [RelaxType.NONE, RelaxType.VOLUME, RelaxType.SHAPE, RelaxType.CELL]:
-            inputs.pw.settings = self._fix_atomic_positions(inputs.pw.structure, inputs.pw.get('settings', None))
-
-        if relax_type in [RelaxType.NONE, RelaxType.ATOMS]:
-            inputs.pw.parameters['CONTROL']['calculation'] = 'relax'
-            inputs.pw.parameters.pop('CELL', None)
-        else:
-            inputs.pw.parameters['CONTROL']['calculation'] = 'vc-relax'
-
-        if relax_type in [RelaxType.VOLUME, RelaxType.ATOMS_VOLUME]:
-            inputs.pw.parameters['CELL']['cell_dofree'] = 'volume'
-
-        if relax_type in [RelaxType.SHAPE, RelaxType.ATOMS_SHAPE]:
-            inputs.pw.parameters['CELL']['cell_dofree'] = 'shape'
-
-        if relax_type in [RelaxType.CELL, RelaxType.ATOMS_CELL]:
-            inputs.pw.parameters['CELL']['cell_dofree'] = 'all'
+        # If the relaxation uses 'vc-relax make sure there is a 'CELL' namelist
+        if inputs.pw.parameters['CONTROL']['calculation'] == 'vc-relax':
+            inputs.pw.parameters.setdefault('CELL', {})
```

Comment:
I think this will now fail because you removed the line `inputs.pw.parameters = inputs.pw.parameters.get_dict()` and so `inputs.pw.parameters` is still a stored `Dict` node and therefore immutable. I don't think we need this anymore do we? Is there any code after this that relies on the `CELL` key being there?

Reply:
Ah, I see now that the unpacking is done in `setup`. Maybe it would be cleaner to move this check there as well if it is still necessary

Reply:
Should we then move _all_ lines that set the parameter inputs to `setup`? I.e. also the three lines above these and those defined in `run_final_scf`?

Reply:
Indeed, it's cleaner if all the input parameter specification is moved to `setup`, I think.